File: frontend/src/api/endpoints/tasks.ts
/**
 * 任务域 hooks（frontend-design.md §5.1 第 3 步 / §6.1 / §8）。
 *
 * 读接口：任务列表分页 + 单任务六类子资源（data/param/metadata/profile/
 * videos/danmakus）；写接口：添加/删除、单任务与批量的启停、录制器开关、
 * 信息刷新。全部经 `call` 拆包 ResponseMessage（§4），mutation 成功后按
 * queryKeys 失效相关缓存（§6.3）。
 *
 * 手动切割（cut）已在后端功能范围决策中移除，无对应端点。
 */
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { api, call } from '../client'
import { queryKeys, type TaskPart, type TasksDataQuery } from '../queryKeys'

/** 列表/详情的实时兜底轮询间隔（毫秒）：速率等运行态字段无 WS 推送。 */
export const TASKS_REFETCH_INTERVAL = 2_000

export interface TasksQueryOptions {
  /** 覆盖轮询间隔；false 关闭轮询（测试用）。 */
  refetchInterval?: number | false
}

/**
 * 任务列表分页数据：`{ total, page, size, tasks }`。
 * `query.select` 由任务筛选器映射（all/living/recording/...，§4）；
 * 运行态速率字段无事件推送，默认以短轮询刷新（§9 兜底策略）。
 */
export function useTasksData(
  query: TasksDataQuery = {},
  options: TasksQueryOptions = {},
) {
  return useQuery({
    queryKey: queryKeys.tasks(query),
    queryFn: () =>
      call(() => api.GET('/api/v1/tasks/data', { params: { query } })),
    refetchInterval: options.refetchInterval ?? TASKS_REFETCH_INTERVAL,
  })
}

/** 单任务子资源 GET 路径映射（六类，§8）。 */
const TASK_PART_PATHS = {
  data: '/api/v1/tasks/{room_id}/data',
  param: '/api/v1/tasks/{room_id}/param',
  metadata: '/api/v1/tasks/{room_id}/metadata',
  profile: '/api/v1/tasks/{room_id}/profile',
  videos: '/api/v1/tasks/{room_id}/videos',
  danmakus: '/api/v1/tasks/{room_id}/danmakus',
} as const satisfies Record<TaskPart, string>

function useTaskPart(
  roomId: number,
  part: TaskPart,
  options: TasksQueryOptions = {},
) {
  return useQuery({
    queryKey: queryKeys.task(roomId, part),
    queryFn: () =>
      call(() =>
        api.GET(TASK_PART_PATHS[part], {
          params: { path: { room_id: roomId } },
        }),
      ),
    refetchInterval: options.refetchInterval ?? false,
  })
}

/** 单任务状态数据（GET /tasks/{room_id}/data），默认短轮询刷新运行态。 */
export function useTaskData(roomId: number, options: TasksQueryOptions = {}) {
  return useTaskPart(roomId, 'data', {
    refetchInterval: options.refetchInterval ?? TASKS_REFETCH_INTERVAL,
  })
}

/** 任务参数快照（GET /tasks/{room_id}/param）。 */
export function useTaskParam(roomId: number) {
  return useTaskPart(roomId, 'param')
}

/** 录制元数据（GET /tasks/{room_id}/metadata）。 */
export function useTaskMetadata(roomId: number) {
  return useTaskPart(roomId, 'metadata')
}

/** 流 ffprobe Profile（GET /tasks/{room_id}/profile）。 */
export function useTaskProfile(roomId: number) {
  return useTaskPart(roomId, 'profile')
}

/** 视频文件明细（GET /tasks/{room_id}/videos）。 */
export function useTaskVideos(roomId: number) {
  return useTaskPart(roomId, 'videos')
}

/** 弹幕文件明细（GET /tasks/{room_id}/danmakus）。 */
export function useTaskDanmakus(roomId: number) {
  return useTaskPart(roomId, 'danmakus')
}

/** mutation 成功后的统一失效：列表前缀 +（可选）单任务全部子资源。 */
function useInvalidateTasks() {
  const queryClient = useQueryClient()
  return async (roomId?: number) => {
    await queryClient.invalidateQueries({ queryKey: ['tasks'] })
    if (roomId !== undefined) {
      await queryClient.invalidateQueries({ queryKey: ['task', roomId] })
    }
  }
}

/**
 * 添加任务（POST /tasks/{room_id}，支持短号，返回真实房号）。
 *
 * 契约缺口：后端未在签名中声明 room_id 路径参数（以 body.room_id 为准），
 * openapi.json 生成 `path?: never`；运行时 openapi-fetch 仍按 params.path
 * 替换 URL 占位符，故此处以 never 断言透传路径参数。契约补全后移除断言。
 */
export function useAddTask() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomId: number) =>
      call(() =>
        api.POST('/api/v1/tasks/{room_id}', {
          params: { path: { room_id: roomId } } as never,
          body: { room_id: roomId, auto_enable: true },
        }),
      ),
    onSuccess: () => invalidate(),
  })
}

/** 启动单任务监控（POST /tasks/{room_id}/start）。 */
export function useStartTask() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomId: number) =>
      call(() =>
        api.POST('/api/v1/tasks/{room_id}/start', {
          params: { path: { room_id: roomId } },
        }),
      ),
    onSuccess: (_data, roomId) => invalidate(roomId),
  })
}

/** 停止单任务监控（POST /tasks/{room_id}/stop）。 */
export function useStopTask() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomId: number) =>
      call(() =>
        api.POST('/api/v1/tasks/{room_id}/stop', {
          params: { path: { room_id: roomId } },
        }),
      ),
    onSuccess: (_data, roomId) => invalidate(roomId),
  })
}

/** 单任务录制器开关（POST /tasks/{room_id}/recorder/enable|disable）。 */
export function useSetTaskRecorder() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: ({ roomId, enabled }: { roomId: number; enabled: boolean }) =>
      call(() =>
        enabled
          ? api.POST('/api/v1/tasks/{room_id}/recorder/enable', {
              params: { path: { room_id: roomId } },
            })
          : api.POST('/api/v1/tasks/{room_id}/recorder/disable', {
              params: { path: { room_id: roomId } },
            }),
      ),
    onSuccess: (_data, { roomId }) => invalidate(roomId),
  })
}

/** 刷新单任务信息（POST /tasks/{room_id}/info）。 */
export function useRefreshTaskInfo() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomId: number) =>
      call(() =>
        api.POST('/api/v1/tasks/{room_id}/info', {
          params: { path: { room_id: roomId } },
        }),
      ),
    onSuccess: (_data, roomId) => invalidate(roomId),
  })
}

/** 删除单任务（DELETE /tasks/{room_id}）。 */
export function useDeleteTask() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomId: number) =>
      call(() =>
        api.DELETE('/api/v1/tasks/{room_id}', {
          params: { path: { room_id: roomId } },
        }),
      ),
    onSuccess: (_data, roomId) => invalidate(roomId),
  })
}

/** 批量启动监控（POST /tasks/start；空数组=全部）。 */
export function useBatchStart() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomIds: number[] = []) =>
      call(() =>
        api.POST('/api/v1/tasks/start', { body: { room_ids: roomIds } }),
      ),
    onSuccess: () => invalidate(),
  })
}

/** 批量停止监控（POST /tasks/stop；空数组=全部）。 */
export function useBatchStop() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomIds: number[] = []) =>
      call(() =>
        api.POST('/api/v1/tasks/stop', {
          body: { room_ids: roomIds, force: false, background: false },
        }),
      ),
    onSuccess: () => invalidate(),
  })
}

/** 批量录制器开关（POST /tasks/recorder/enable|disable；空数组=全部）。 */
export function useBatchSetRecorder() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: ({
      roomIds = [],
      enabled,
    }: {
      roomIds?: number[]
      enabled: boolean
    }) =>
      call(() =>
        enabled
          ? api.POST('/api/v1/tasks/recorder/enable', {
              body: { room_ids: roomIds },
            })
          : api.POST('/api/v1/tasks/recorder/disable', {
              body: { room_ids: roomIds },
            }),
      ),
    onSuccess: () => invalidate(),
  })
}

/** 批量删除任务（DELETE /tasks）。 */
export function useDeleteTasks() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: (roomIds: number[]) =>
      call(() => api.DELETE('/api/v1/tasks', { body: { room_ids: roomIds } })),
    onSuccess: () => invalidate(),
  })
}

/** 批量刷新任务信息（POST /tasks/info）。 */
export function useBatchRefreshInfo() {
  const invalidate = useInvalidateTasks()
  return useMutation({
    mutationFn: () => call(() => api.POST('/api/v1/tasks/info')),
    onSuccess: () => invalidate(),
  })
}

File: src/birec/web/routers/tasks.py
"""Task API endpoints (§7.1).

Provides CRUD and control operations for record tasks, wired to the
``RecordTaskManager`` stored in ``app.state.task_manager``.

Route ordering: fixed-path batch endpoints are registered BEFORE
``/{room_id}`` parameterised routes so FastAPI matches them correctly.
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Any

from fastapi import APIRouter, Query, Request
from pydantic import BaseModel

from ...bili.exceptions import ApiRequestError
from ...task import RecordTaskManager, RunningStatus, TaskData
from ..models import ResponseMessage

__all__ = ("router",)

router = APIRouter(prefix="/api/v1/tasks", tags=["tasks"])


# ── Request models ────────────────────────────────────────────────────────────


class AddTaskRequest(BaseModel):
    """Request body for adding a task.

    The room is always written to the config file, so the task survives a
    restart and has task-level options to patch. ``auto_enable`` decides
    whether monitoring and recording start enabled (§5.2).
    """

    room_id: int
    auto_enable: bool = True


class BatchRoomIds(BaseModel):
    """Request body for batch operations on specific rooms."""

    room_ids: list[int] = []


class BatchStopRequest(BaseModel):
    """Request body for batch stop with options."""

    room_ids: list[int] = []
    force: bool = False
    background: bool = False


class DeleteTasksRequest(BaseModel):
    """Request body for batch delete."""

    room_ids: list[int]


# ── Helpers ───────────────────────────────────────────────────────────────────


def _get_manager(request: Request) -> RecordTaskManager:
    """Extract the RecordTaskManager from app state."""
    manager: RecordTaskManager = request.app.state.task_manager
    return manager


def _ok(data: dict[str, Any] | None = None, message: str = "") -> dict[str, Any]:
    return ResponseMessage(code=0, message=message, data=data).to_dict()


def _task_data_to_dict(td: TaskData) -> dict[str, Any]:
    """Convert TaskData to a JSON-serialisable dict."""
    d = asdict(td)
    # Enum → value
    d["task_status"]["running_status"] = td.task_status.running_status.value
    return d


def _filter_tasks(
    tasks: list[TaskData],
    select: str | None,
) -> list[TaskData]:
    """Filter task data list by the ``select`` query parameter."""
    if not select or select == "all":
        return tasks

    result: list[TaskData] = []
    for td in tasks:
        if _matches_filter(td, select):
            result.append(td)
    return result


def _matches_filter(td: TaskData, select: str) -> bool:
    """Check if a task matches the given filter."""
    if select == "preparing":
        return not td.live_status
    if select == "living":
        return td.live_status
    if select == "monitor_enabled":
        return td.task_status.monitor_enabled
    if select == "monitor_disabled":
        return not td.task_status.monitor_enabled
    if select == "recorder_enabled":
        return td.task_status.recorder_enabled
    if select == "recorder_disabled":
        return not td.task_status.recorder_enabled
    if select in ("stopped", "waiting", "recording", "remuxing", "injecting"):
        return td.task_status.running_status == RunningStatus(select)
    return False


# ── GET /tasks/data (fixed path, before /{room_id}) ──────────────────────────


@router.get("/data")
async def get_tasks_data(
    request: Request,
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    select: str | None = Query(None),
) -> dict[str, Any]:
    """Get paginated task data with optional filtering."""
    manager = _get_manager(request)
    all_data = manager.get_all_task_data()
    filtered = _filter_tasks(all_data, select)
    total = len(filtered)
    start = (page - 1) * size
    end = start + size
    page_data = filtered[start:end]
    return _ok(
        {
            "total": total,
            "page": page,
            "size": size,
            "tasks": [_task_data_to_dict(td) for td in page_data],
        }
    )


# ── POST batch endpoints (fixed paths, BEFORE /{room_id}) ────────────────────


@router.post("/info")
async def batch_refresh_info(request: Request) -> dict[str, Any]:
    """Refresh info for all tasks."""
    manager = _get_manager(request)
    await manager.batch_refresh_info()
    return _ok(message="Info refreshed")


@router.post("/start")
async def batch_start(
    request: Request, body: BatchRoomIds | None = None
) -> dict[str, Any]:
    """Start monitoring for specified or all tasks."""
    manager = _get_manager(request)
    if body and body.room_ids:
        for rid in body.room_ids:
            task = manager.get_task(rid)
            if task:
                await task.enable_monitor()
    else:
        for task in manager.get_all_tasks():
            await task.enable_monitor()
    return _ok(message="Started")


@router.post("/stop")
async def batch_stop(
    request: Request, body: BatchStopRequest | None = None
) -> dict[str, Any]:
    """Stop monitoring for specified or all tasks."""
    manager = _get_manager(request)
    if body and body.room_ids:
        for rid in body.room_ids:
            task = manager.get_task(rid)
            if task:
                await task.disable_monitor()
    else:
        for task in manager.get_all_tasks():
            await task.disable_monitor()
    return _ok(message="Stopped")


@router.post("/recorder/enable")
async def batch_enable_recorder(
    request: Request, body: BatchRoomIds | None = None
) -> dict[str, Any]:
    """Enable recorder for specified or all tasks."""
    manager = _get_manager(request)
    if body and body.room_ids:
        for rid in body.room_ids:
            task = manager.get_task(rid)
            if task:
                task.enable_recorder()
    else:
        for task in manager.get_all_tasks():
            task.enable_recorder()
    return _ok(message="Recorder enabled")


@router.post("/recorder/disable")
async def batch_disable_recorder(
    request: Request, body: BatchRoomIds | None = None
) -> dict[str, Any]:
    """Disable recorder for specified or all tasks."""
    manager = _get_manager(request)
    if body and body.room_ids:
        for rid in body.room_ids:
            task = manager.get_task(rid)
            if task:
                await task.disable_recorder()
    else:
        for task in manager.get_all_tasks():
            await task.disable_recorder()
    return _ok(message="Recorder disabled")


# ── DELETE batch (fixed path) ────────────────────────────────────────────────


@router.delete("")
async def delete_tasks(request: Request, body: DeleteTasksRequest) -> dict[str, Any]:
    """Delete multiple tasks."""
    manager = _get_manager(request)
    errors: list[str] = []
    for rid in body.room_ids:
        try:
            await manager.remove_task(rid)
        except KeyError:
            errors.append(f"Task {rid} not found")
    if errors:
        return ResponseMessage(code=404, message="; ".join(errors)).to_dict()
    return _ok(message="Deleted")


# ── GET /{room_id}/* endpoints ───────────────────────────────────────────────


@router.get("/{room_id}/data")
async def get_task_data(request: Request, room_id: int) -> dict[str, Any]:
    """Get data for a single task."""
    manager = _get_manager(request)
    try:
        td = manager.get_task_data(room_id)
    except KeyError:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    return _ok(_task_data_to_dict(td))


@router.get("/{room_id}/param")
async def get_task_param(request: Request, room_id: int) -> dict[str, Any]:
    """Get task parameters."""
    manager = _get_manager(request)
    task = manager.get_task(room_id)
    if task is None:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    param = task.get_param()
    return _ok(asdict(param))


@router.get("/{room_id}/metadata")
async def get_task_metadata(request: Request, room_id: int) -> dict[str, Any]:
    """Get recording metadata for a task."""
    manager = _get_manager(request)
    task = manager.get_task(room_id)
    if task is None:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    metadata = task.get_metadata()
    return _ok(asdict(metadata))


@router.get("/{room_id}/profile")
async def get_task_profile(request: Request, room_id: int) -> dict[str, Any]:
    """Get current stream ffprobe profile."""
    manager = _get_manager(request)
    task = manager.get_task(room_id)
    if task is None:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    profile = task.get_profile()
    return _ok(profile)


@router.get("/{room_id}/videos")
async def get_task_videos(request: Request, room_id: int) -> dict[str, Any]:
    """Get video file details for a task."""
    manager = _get_manager(request)
    task = manager.get_task(room_id)
    if task is None:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    videos = task.get_videos()
    return _ok({"videos": [asdict(v) for v in videos]})


@router.get("/{room_id}/danmakus")
async def get_task_danmakus(request: Request, room_id: int) -> dict[str, Any]:
    """Get danmaku file details for a task."""
    manager = _get_manager(request)
    task = manager.get_task(room_id)
    if task is None:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    danmakus = task.get_danmakus()
    return _ok({"danmakus": [asdict(d) for d in danmakus]})


# ── POST /{room_id}/* endpoints ──────────────────────────────────────────────


@router.post("/{room_id}/info")
async def refresh_task_info(request: Request, room_id: int) -> dict[str, Any]:
    """Refresh info for a single task."""
    manager = _get_manager(request)
    task = manager.get_task(room_id)
    if task is None:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    await task.refresh_info()
    return _ok(message="Info refreshed")


@router.post("/{room_id}/start")
async def start_task(request: Request, room_id: int) -> dict[str, Any]:
    """Start monitoring for a task."""
    manager = _get_manager(request)
    try:
        await manager.start_task(room_id)
    except KeyError:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    return _ok(message="Started")


@router.post("/{room_id}/stop")
async def stop_task(request: Request, room_id: int) -> dict[str, Any]:
    """Stop monitoring for a task."""
    manager = _get_manager(request)
    try:
        await manager.stop_task(room_id)
    except KeyError:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    return _ok(message="Stopped")


@router.post("/{room_id}/recorder/enable")
async def enable_recorder(request: Request, room_id: int) -> dict[str, Any]:
    """Enable recorder for a task."""
    manager = _get_manager(request)
    try:
        manager.enable_recorder(room_id)
    except KeyError:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    return _ok(message="Recorder enabled")


@router.post("/{room_id}/recorder/disable")
async def disable_recorder(request: Request, room_id: int) -> dict[str, Any]:
    """Disable recorder for a task."""
    manager = _get_manager(request)
    try:
        await manager.disable_recorder(room_id)
    except KeyError:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    return _ok(message="Recorder disabled")


# ── POST /{room_id} (add task — MUST be after all fixed POST paths) ──────────


@router.post("/{room_id}")
async def add_task(request: Request, body: AddTaskRequest) -> dict[str, Any]:
    """Add a new task for a room."""
    manager = _get_manager(request)
    try:
        task = await manager.add_task(body.room_id, auto_enable=body.auto_enable)
    except ValueError as exc:
        return ResponseMessage(code=409, message=str(exc)).to_dict()
    except ApiRequestError as exc:
        # Bilibili refused to tell us about the room. Without its info there is
        # no task to build, and the caller needs to hear that as an answer
        # rather than as a 500 from a stray exception.
        return ResponseMessage(
            code=502, message=f"Bilibili API error: {exc.message or exc.code}"
        ).to_dict()
    except RuntimeError as exc:
        return ResponseMessage(code=500, message=str(exc)).to_dict()
    return _ok({"room_id": task.room_id}, message="Task added")


# ── DELETE /{room_id} ────────────────────────────────────────────────────────


@router.delete("/{room_id}")
async def delete_task(request: Request, room_id: int) -> dict[str, Any]:
    """Delete a single task."""
    manager = _get_manager(request)
    try:
        await manager.remove_task(room_id)
    except KeyError:
        return ResponseMessage(code=404, message=f"Task {room_id} not found").to_dict()
    return _ok(message="Deleted")
